src/pymor/reductors/neural_network.py

```diff
@@ -0,0 +1,459 @@
+                    for phase in phases:
+                        if phase == 'train':
+                            neural_network.train()
+                        else:
+                            neural_network.eval()
+
+                        running_loss = 0.0
+
+                        # iterate over batches
+                        for batch in dataloaders[phase]:
+                            inputs = batch[0]
+                            targets = batch[1]
+
+                            with torch.set_grad_enabled(phase == 'train'):
+                                def closure():
+                                    if torch.is_grad_enabled():
+                                        optimizer.zero_grad()
+                                    outputs = neural_network(inputs)
+                                    loss = loss_function(outputs, targets)
+                                    if loss.requires_grad:
+                                        loss.backward()
+                                    return loss
+
+                                # perform optimization step
+                                if phase == 'train':
+                                    optimizer.step(closure)
+
+                                # compute loss of current batch
+                                loss = closure()
+
+                            # update overall absolute loss
+                            running_loss += loss.item() * len(batch[0])
+
+                        # compute average loss
+                        epoch_loss = running_loss / len(dataloaders[phase].dataset)
+
+                        losses[phase] = epoch_loss
+
+                        # check for early stopping
+                        if phase == 'val' and early_stopping_scheduler(losses, neural_network):
+                            if not self.logging_disabled:
+                                self.logger.info(f'Early stopping training process after {epoch + 1} epochs ...')
+                                self.logger.info(f'Minimum validation loss: {early_stopping_scheduler.best_losses["val"]}')
+                            return early_stopping_scheduler.best_neural_network, early_stopping_scheduler.best_losses
+
+            return early_stopping_scheduler.best_neural_network, early_stopping_scheduler.best_losses
+
+        def build_basis(self):
+            """Compute a reduced basis using proper orthogonal decomposition."""
+            snapshots = []
+            self.training_data = []
+
+            with self.logger.block('Building reduced basis ...'):
+
+                # compute snapshots for POD and training of neural networks
+                with self.logger.block('Computing training snapshots ...'):
+                    U = self.fom.solution_space.empty()
+                    for mu in self.training_set:
+                        u = self.fom.solve(mu)
+                        U.append(u)
+                        snapshots.append({'mu': mu, 'u_full': u})
```

Comment:
For at least `NumpyVectorArray` this makes you keep the entire snapshot data twice in memory which may be infeasible. Also I do not think that you need `snapshots`. Just use `U` and `training_set`.

Reply:
Your right, `U` and `training_set` should be indeed sufficient.

Reply:
I thought a bit more about this and from my point of view, we do not need `U` after constructing the reduced basis and performing the projection of the snapshots in `U` onto the reduced basis. In `training_data`, we store the parameters (which indeed could somehow be derived from `training_set`) and the coefficients of the projections of the full-order snapshots onto the reduced basis. This seems to be much cheaper than storing `U` the whole time and always projecting onto the reduced space over and over again.

Reply:
Alright, I think now I got your idea and I changed this part accordingly. Hope this is as you want it?

Reply:
Now the changes should be there again.